### PR TITLE
add ILanguageServerInfoProvider interface

### PR DESCRIPTION
**Pull request by ansyral**
@andxu @testforstephen @yaohaizh @akaroml

### CODE REVIEW

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ILanguageServerInfoProvider.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+public interface ILanguageServerInfoProvider {
+    String getFullyQualifiedName(String sourceFilePath, int line, int column);
+
+    String getSourceFilePath(String fullyQualifiedName, String projectName);
+
+    String getURI(String fullyQualifiedName, String projectName);
+
+    String getSourceContents(String uri);
+
+    String[] getClassPath(String mainClassFullyQualifiedName, String projectName);
```
Comment (andxu) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ILanguageServerInfoProvider.java]: why we need mainClassFullyQualifiedName/projectName to getClassPath

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ILanguageServerInfoProvider.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package org.eclipse.jdt.ls.debug.adapter;
+
+public interface ILanguageServerInfoProvider {
+    String getFullyQualifiedName(String sourceFilePath, int line, int column);
```
Comment (testforstephen) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ILanguageServerInfoProvider.java]: we don't support column breakpoint. so here just need two parameters `sourceFilePath` and `line`

Comment (testforstephen) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ILanguageServerInfoProvider.java]: please also add batch api `String[] getFullyQualifiedName(String sourceFilePath, int[] lines)`

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ILanguageServerInfoProvider.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+    String getSourceFilePath(String fullyQualifiedName, String projectName);
+
+    String getURI(String fullyQualifiedName, String projectName);
```
Comment (testforstephen) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ILanguageServerInfoProvider.java]: Actually `getSourceFilePath` and `getURI` do the same thing. Just need one API `getURI`.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ILanguageServerInfoProvider.java (hunk 2 above)
Comment (andxu) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ILanguageServerInfoProvider.java]: String[] getFullyQualifiedName(String sourceFilePath, int[] lines) is enough.

Comment (ansyral) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ILanguageServerInfoProvider.java]: How about the caller to do the batch operation as each line only resolve one fullyqualifiedname?

Comment (ansyral) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ILanguageServerInfoProvider.java]: the method isn't only for getting breakpoint?

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IVirtualMachineManagerProvider.java [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package org.eclipse.jdt.ls.debug.adapter;
+
+public interface IVirtualMachineManagerProvider {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IVirtualMachineManagerProvider.java]: What's this for?

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ISourceLookUpProvider.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package org.eclipse.jdt.ls.debug.adapter;
+
+import java.util.Map;
+
+public interface ISourceLookUpProvider {
+    void initializeContext(Map<String, Object> props);
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ISourceLookUpProvider.java]: If all provider shares the same initialize routine, make it another interface and let others extend it.

**Review by akaroml — CHANGES_REQUESTED** [review]

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IVirtualMachineManagerProvider.java (hunk 4 above)
Comment (ansyral) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IVirtualMachineManagerProvider.java]: current implementation is that debugger used the same virtualmachinemanager as language server, which is `org.eclipse.jdi.Bootstrap.virtualMachineManager`, we can encapsulate this.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ISourceLookUpProvider.java (hunk 5 above)
Comment (ansyral) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ISourceLookUpProvider.java]: cool

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IHasContext.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package org.eclipse.jdt.ls.debug.adapter;
+
+import java.util.Map;
+
+public interface IHasContext {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IHasContext.java]: Not quite sure about the naming. Is the context used by providers? If that's the case, then it's a common behavior across all providers. And a common IProvider is recommended.

**Review by akaroml — CHANGES_REQUESTED** [review]

**Review by akaroml — APPROVED** [review]

**Comment by akaroml** [comment]
Please fix the build error.

**Pull request merged**